packages/SwingSet/src/kernel/vatManager/snapStore.js

```diff
@@ -0,0 +1,127 @@
+// @ts-check
+import { createHash } from 'crypto';
+import { pipeline } from 'stream';
+import { createGzip, createGunzip } from 'zlib';
+import { assert, details as d } from '@agoric/assert';
+import { promisify } from 'util';
+
+const pipe = promisify(pipeline);
+
+const { freeze } = Object;
+
+/**
+ * @param {string} root
+ * @param {{
+ *   tmpName: typeof import('tmp').tmpName,
+ *   existsSync: typeof import('fs').existsSync
+ *   createReadStream: typeof import('fs').createReadStream,
+ *   createWriteStream: typeof import('fs').createWriteStream,
+ *   resolve: typeof import('path').resolve,
+ *   rename: typeof import('fs').promises.rename,
+ *   unlink: typeof import('fs').promises.unlink,
+ * }} io
+ */
+export function makeSnapstore(
+  root,
+  {
+    tmpName,
+    existsSync,
+    createReadStream,
+    createWriteStream,
+    resolve,
+    rename,
+    unlink,
+  },
+) {
+  /** @type {(opts: unknown) => Promise<string>} */
+  const ptmpName = promisify(tmpName);
+  const tmpOpts = { tmpdir: root, template: 'tmp-XXXXXX.xss' };
+  /**
+   * @param { (name: string) => Promise<T> } thunk
+   * @returns { Promise<T> }
+   * @template T
+   */
+  async function withTempName(thunk) {
+    const name = await ptmpName(tmpOpts);
+    const result = await thunk(name);
```

Comment:
If the thunk throws, does this leave the tempfile hanging around? Should the thunk be called inside the `try`, and the `unlink` in a `finally`?